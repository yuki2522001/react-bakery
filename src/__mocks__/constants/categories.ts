import { CategoryProps } from "@common-types/category";

const CATEGORY_MOCKING_LIST: CategoryProps[] = [
  {
    id: "1",
    name: "Cake"
  },
  {
    id: "2",
    name: "Breads"
  }
]

export { CATEGORY_MOCKING_LIST };
