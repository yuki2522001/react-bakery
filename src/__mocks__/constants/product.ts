import { Product } from "@common-types/product";
import url from "../../assets/images/products/black-forest-cake.jpg";

const PRODUCT_MOCKING: Product = {
  id: "1",
  categoryId: "1651924356334",
  name: "Cheese cake",
  price: 20,
  quantity: 5,
  description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
  images: [
    url
  ]
}

const PRODUCT_MOCKING_LIST: Product[] = [
  {
    id: "1",
    categoryId: "1651924356334",
    name: "Cheese cake",
    price: 20,
    quantity: 5,
    description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
    images: [
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg"
    ]
  },
  {
    id: "2",
    categoryId: "1651924356334",
    name: "Amaretti Cookie",
    price: 10,
    quantity: 5,
    description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
    images: [
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg"
    ]
  }
]

export { PRODUCT_MOCKING, PRODUCT_MOCKING_LIST };
