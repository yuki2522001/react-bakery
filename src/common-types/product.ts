import { CategoryProps } from "./category";

export interface Product {
  id: string;
  name: string;
  price: number;
  images: string[];
  quantity: number;
  categoryId: string;
  description: string;
  product?: {};
}

export interface ProductContext {
  products: Product[];
  dispatch: Function,
  searchValue: string;
  setSearchValue: Function;
  categories: CategoryProps[];
  setCategories: Function;
}