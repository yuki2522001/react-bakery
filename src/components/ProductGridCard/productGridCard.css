.product {
  position: relative;
  margin: 18px;
  z-index: 0;
}

.product,
.product__image {
  width: 250px;
}

.product__image:hover {
  opacity: 0.5;
}

.product__image:hover+.btn__delete {
  display: block;
}

.productViewPage__link {
  color: var(--text-master);
  text-decoration: none;
}

.productViewPage__title {
  cursor: pointer;
  font-size: 16px;
}

.product__content {
  text-align: center;
}

.product__title {
  font-size: 16px;
  cursor: pointer;
  text-transform: uppercase;
}

.productViewPage__price {
  color: var(--text-primary);
  font-weight: bold;
  margin: 0;
}