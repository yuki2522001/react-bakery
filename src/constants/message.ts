const SUCCESS_MSG = {
  MESSAGE_ADD_PRODUCT: "Product add successful",
  MESSAGE_UPDATE_PRODUCT: "Product update successful",
  MESSAGE_DELETE_PRODUCT: "Product delete successful",
};

const ERROR_MSG = {
  REQUIRED: "Field is required",
  NUMBER: "Enter a non-negative number"
}

export { SUCCESS_MSG, ERROR_MSG };
