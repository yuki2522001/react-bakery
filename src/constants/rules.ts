const RULES = {
  REQUIRED: "required",
  NUMBER: "number",
  NEGATIVE: "negative"
};

export { RULES }