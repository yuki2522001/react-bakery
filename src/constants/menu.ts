import { MenuType } from "../common-types/menu";

const MENU_LIST: MenuType[] = [
  {
    key: "about",
    label: "About",
  },
  {
    key: "shop",
    label: "Shop",
  },
  {
    key: "blog",
    label: "Blog",
  },
  {
    key: "faq",
    label: "FAQ",
  },
  {
    key: "contact",
    label: "Contact"
  }
];

export default MENU_LIST;
