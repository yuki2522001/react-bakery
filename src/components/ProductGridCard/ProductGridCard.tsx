import React, { useCallback, useState } from "react";
import { Link } from "react-router-dom";
import { Product } from "@common-types/product";
import Title from "@components/common/Title/Title";
import Price from "@components/Price/Price";
import ModalDelete from "@components/Modal/ModalDelete/ModalDelete";
import "./productGridCard.css";

interface ProductItemProps {
  product: Product;
  deleteProduct: (id: string) => void;
}

const ProductGridCard: React.FC<ProductItemProps> = ({
  product,
  deleteProduct,
}) => {
  const [openModalDelete, setOpenModalDelete] = useState<boolean>(false);
  // handle toggle modal delete
  const toggleModalDelete = useCallback(() => {
    setOpenModalDelete(!openModalDelete);
  }, [openModalDelete]);

  return (
    <>
      <div data-testid="product-grid-card" className="product">
        <img className="product__image" src={product.images[0]} />
        <button
          data-testid="open-modal-delete"
          onClick={toggleModalDelete}
          className="btn btn__delete"
        >
          <i className="fa fa-trash"></i>
        </button>
        <div className="product__content">
          <Link className="productViewPage__link" to={`/products/${product.id}`}>
            <Title className="productViewPage__title" text={product.name} />
          </Link>
          <Price className="productViewPage__price" value={product.price} />
        </div>
        {openModalDelete && (
          <ModalDelete
            id={product.id}
            deleteProduct={deleteProduct}
            hideModalDelete={toggleModalDelete}
          />
        )}
      </div>
    </>
  );
};

export default ProductGridCard;
