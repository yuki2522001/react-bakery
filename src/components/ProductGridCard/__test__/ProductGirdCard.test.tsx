import { fireEvent, render } from "@testing-library/react";
import "@testing-library/jest-dom";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import ProductGridCard from "../ProductGridCard";
import { useState } from "react";
import { PRODUCT_MOCKING } from "@__mocks__/constants/product";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

describe("Product grid card component", () => {
  beforeEach(()=>{
    (useState as jest.Mock).mockImplementation(jest.requireActual("react").useState);
  })

  test("should render product grid card component", () => {
    const history = createMemoryHistory();
    const { getByTestId } = render (
      <Router location={history.location} navigator={history}>
        <ProductGridCard product={PRODUCT_MOCKING} deleteProduct={() => {}} />
      </Router>
    );
    expect(getByTestId("product-grid-card")).toBeInTheDocument();
  });

  test("should open modal delete when click button delete", () => {
    const history = createMemoryHistory();
    const { getByTestId } = render (
      <Router location={history.location} navigator={history}>
        <ProductGridCard product={PRODUCT_MOCKING} deleteProduct={() => {}} />
      </Router>
    );
    const btnOpenModal = getByTestId("open-modal-delete");
    fireEvent.click(btnOpenModal);
    expect(btnOpenModal).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Router location={history.location} navigator={history}>
        <ProductGridCard product={PRODUCT_MOCKING} deleteProduct={() => {}} />
      </Router>);
    expect(asFragment()).toMatchSnapshot();
  });
})