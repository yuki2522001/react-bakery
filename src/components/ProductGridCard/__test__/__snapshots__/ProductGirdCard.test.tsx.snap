// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product grid card component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="product"
    data-testid="product-grid-card"
  >
    <img
      class="product__image"
    />
    <button
      class="btn btn__delete"
      data-testid="open-modal-delete"
    >
      <i
        class="fa fa-trash"
      />
    </button>
    <div
      class="product__content"
    >
      <a
        class="productViewPage__link"
        href="/products/1"
      >
        <h2
          class="productViewPage__title"
          data-testid="title-test"
        >
          Cheese cake
        </h2>
      </a>
      <p
        class="productViewPage__price"
        data-testid="price"
      >
        $20
      </p>
    </div>
  </div>
</DocumentFragment>
`;
